### PR TITLE
Add tag command

### PR DESCRIPTION
**Pull request by ncdc**

### CODE REVIEW

**Comment by ncdc** [comment]
@smarterclayton PTAL. I still have some TODOs and I haven't updated the tests to fully compile. But let me know how this is looking. I got rid of the docker flag - if we can't find the ISTag or ISImage, we treat it as a Docker ref.

**Comment by liggitt** [comment]
> if we can't find the ISTag or ISImage, we treat it as a Docker ref

do we still want a way to unambiguously state the type?

**Comment by smarterclayton** [comment]
docker:foo/bar maybe

**Comment by smarterclayton** [comment]
Will review tonight

----- Original Message -----

**Comment by ncdc** [comment]
> do we still want a way to unambiguously state the type? magic is awesome, but so are unambiguous commands (in the case where the user knows exactly what they want)

I can add a --source flag with valid options docker, imagestreamtag aka istag, imagestreamimage aka isimage.

**Comment by smarterclayton** [comment]
How common is the cross namespace tag?  I'm wondering whether we should just use straight up resource type notation here.

**Comment by ncdc** [comment]
If anyone wants to tag 1 of the stock OpenShift builder images into their own streams, I'd say pretty common.

What do you mean by resource type notation?

**Comment by smarterclayton** [comment]
istag/foo:bar

**Comment by ncdc** [comment]
That's what I thought. We could do that but it's not as pretty.

**Comment by smarterclayton** [comment]
The problem is that it's standard.  It's the path of the tools.  It's just not easy to cross namespace tag

**Comment by ncdc** [comment]
Yeah. Options I can think of:

1. `oc tag -n openshift istag/ruby:2.0 ...` this doesn't seem right, as the target streams would default to the source namespace (openshift)
1. `oc tag istag/openshift/ruby:2.0 ...` this looks weird
1. `oc tag openshift/ruby:2.0 ...` (assume istag, fall back to docker pull spec)
1. `oc tag --source=istag openshift/ruby:2.0 ...` (make source type explicit)

I still like 3 and 4. Any other options I missed?

**Comment by smarterclayton** [comment]
> On Jun 8, 2015, at 8:55 PM, Andy Goldstein <notifications@github.com> wrote:
> 
> Yeah. Options I can think of:
> 
> oc tag -n openshift istag/ruby:2.0 ... this doesn't seem right, as the target streams would default to the source namespace (openshift)
> oc tag istag/openshift/ruby:2.0 ... this looks weird
We'll eventually do namespace/istag/name probably.
> oc tag openshift/ruby:2.0 ... (assume istag, fall back to docker pull spec)
I think a docker image is the exception, not istag.  So --from=docker or --from=registry.  I guess I'm ok with 3 more than 4.  Need to doc this well.  Destinations can only be istags.

**Comment by ncdc** [comment]
I agree that docker image is the exception. 3 and 4 are the same in my book - 4 just lets you specify the source type explicitly.

Do you want me to do namespace/type/name now, or later?

**Comment by smarterclayton** [comment]
> On Jun 8, 2015, at 9:14 PM, Andy Goldstein <notifications@github.com> wrote:
> 
> I agree that docker image is the exception. 3 and 4 are the same in my book - 4 just lets you specify the source type explicitly.
> 
> Do you want me to do namespace/type/name now, or later?
> 
It's a later thing.  Let's do 3

**Comment by ncdc** [comment]
@smarterclayton PR updated, PTAL

File: pkg/cmd/cli/cmd/tag.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	cmdutil "github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/cmd/util"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+	"github.com/openshift/origin/pkg/image/registry/imagestreamimage"
+	"github.com/spf13/cobra"
+
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+)
+
+const (
+	tagLong = `Tag an image into 1 or more image streams.
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/tag.go]: No period at the end of the sentence, "Tag images into image streams"

File: pkg/cmd/cli/cmd/tag.go [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  $ %[1]s tag --source=docker openshift/origin:latest yourproject/ruby:tip`
+)
+
+// NewCmdTag implements the OpenShift cli tag command.
+func NewCmdTag(fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	var sourceKind string
+
+	cmd := &cobra.Command{
+		Use:     "tag [--source=SOURCETYPE] SOURCE DEST [DEST ...]",
+		Short:   "Tag an image into 1 or more image streams.",
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/tag.go]: "Tag existing images into image streams"

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/tag.go]: No period

File: pkg/cmd/cli/cmd/tag.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+		Long:    tagLong,
+		Example: fmt.Sprintf(tagExample, fullName),
+		Run: func(cmd *cobra.Command, args []string) {
+			resolvedSourceKind := ""
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/tag.go]: Move this into RunTag

File: pkg/cmd/cli/cmd/tag.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+		return "", "", fmt.Errorf("invalid name %q for image stream %q", streamName, name)
+	}
+
+	return namespace, streamName, nil
+}
+
+// RunTag contains all the necessary functionality for the OpenShift cli tag command.
+func RunTag(f *clientcmd.Factory, out io.Writer, cmd *cobra.Command, args []string, sourceKind string) error {
+	if len(args) < 2 {
+		return cmdutil.UsageError(cmd, "you must specify a source and at least 1 destination.")
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/tag.go]: "one" not "1"

**Comment by ncdc** [comment]
Comments addressed

File: pkg/cmd/cli/cmd/tag.go [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+	}
+
+	cmd.Flags().StringVar(&sourceKind, "source", sourceKind, "Optional hint for the source type; valid values are 'imagestreamtag', 'istag', 'imagestreamimage', 'isimage', and 'docker'")
+
+	return cmd
+}
+
+func determineSourceKind(input string) string {
+	switch strings.ToLower(input) {
+	case "istag", "imagestreamtag":
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/tag.go]: You should be using RESTMapper for this.

Comment (ncdc) [inline pkg/cmd/cli/cmd/tag.go]: That won't cover DockerImage, will it?

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/tag.go]: Add something to rest mapper or just have an else block in here after you ask it.

File: pkg/cmd/cli/cmd/tag.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+		if err != nil {
+			return err
+		}
+
+		destName, destTag, ok := imageapi.SplitImageStreamTag(destNameAndTag)
+		if !ok {
+			return fmt.Errorf("%q must be of the form <namespace>/<stream name>:<tag>", arg)
+		}
+
+		destClient := osClient.ImageStreams(destNamespace)
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/tag.go]: isc or something shorter

File: pkg/cmd/cli/cmd/tag.go [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+		destStream, err := destClient.Get(destName)
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/tag.go]: target

File: pkg/cmd/cli/cmd/tag.go [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+		if err != nil {
+			return err
+		}
+
+		if destStream.Spec.Tags == nil {
+			destStream.Spec.Tags = make(map[string]imageapi.TagReference)
+		}
+
+		tagRef, ok := destStream.Spec.Tags[destTag]
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/tag.go]: targetRef

**Comment by smarterclayton** [comment]
A few minor comments, then LGTM

**Comment by smarterclayton** [comment]
Approved

**Comment by ncdc** [comment]
@smarterclayton PTAL at the RESTMapper changes

File: pkg/cmd/cli/cmd/tag.go [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+			sourceKind = "DockerImage"
+		}
+	}
+
+	// try RESTMapper if it's not DockerImage
+	if len(sourceKind) > 0 && sourceKind != "DockerImage" {
+		original := sourceKind
+
+		mapper, _ := f.Object()
+		_, kind, _ := mapper.VersionAndKindForResource(sourceKind)
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/tag.go]: Do this the other way around - just call mapper, if you get an error it's not recognized, then check docker image.

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/tag.go]: Then check valid sources.  Also, put this in a method.

Comment (ncdc) [inline pkg/cmd/cli/cmd/tag.go]: How about now?

**Comment by smarterclayton** [comment]
Add a hack/test-cmd.sh and you can get merged

**Comment by ncdc** [comment]
[test]

**Comment by smarterclayton** [comment]
LGTM [merge]

**Pull request merged**